src/inmanta/ast/statements/__init__.py

```diff
@@ -127,7 +127,7 @@ def execute(self, requires: Dict[object, object], resolver: Resolver, queue: Que
         raise NotImplementedError()
 
     def requires_emit_gradual(
-        self, resolver: Resolver, queue: QueueScheduler, resultcollector: ResultCollector
+        self, resolver: Resolver, queue: QueueScheduler, resultcollector: Optional[ResultCollector]
```

Comment:
Isn't the presence of a result collector essential for the principle of gradual emit?

Reply:
I see that `CreateList` allows it to be optional for some reason but
1. that seems like an exception so why not only broaden the type there?
2. even there I'm not sure what purpose it serves. It seems there was a bug there so is the path even reachable?

Reply:
The result collector is indeed expected here, but in many places it is convenient to make `requires_emit` just call `requires_emit_gradual` with no result collector.  

Widening it in one place would violate lyskov substitution. 

The choice was to either 
1. make it non-optional everywhere and have every call to `requires_emit_gradual` perform a null check (or at least reform this to never have nulls)
2. make it optional everywhere and make all  `requires_emit_gradual` methods that are null sensitive delegate to `requires_emit`

Reply:
Are you sure widening would violate liskov? Isn't it the other way around? You can safely widen a function parameter and safely narrow a return value, can't you?

Reply:
ooh, right.

Reply:
(dang, I got that all twisted up in my head, since a looong time)

fixed